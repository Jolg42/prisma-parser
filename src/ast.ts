export type Node = DocumentNode | ModelDefinitionNode | FieldDefinitionNode | TypeNode | IdentifierNode | AttributeNode

export type Position = {
  line: number
  column: number
  offset: number
}

export type Location = {
  start: Position
  end: Position
}

export type DocumentNode = {
  kind: 'Document'
  definitions: ModelDefinitionNode[]
  location?: Location
}

export type ModelDefinitionNode = {
  kind: 'ModelDefinition'
  name: IdentifierNode
  fields: FieldDefinitionNode[]
  location?: Location
}

export type FieldDefinitionNode = {
  kind: 'FieldDefinition'
  name: IdentifierNode
  type: TypeNode
  attributes: AttributeNode[]
  location?: Location
}

export type TypeModifier = 'none' | 'array' | 'optional'

export type TypeNode = {
  kind: 'Type'
  name: IdentifierNode
  modifier: TypeModifier
  location?: Location
}

export type IdentifierNode = {
  kind: 'Identifier'
  identifier: string
  location?: Location
}

export type AttributeNode = {
  kind: 'Attribute'
  name: string
  location?: Location
}
