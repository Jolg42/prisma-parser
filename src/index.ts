export { parse } from './parser'
export { print } from './printer'
export { PrismaParsingError } from './prisma-parsing-error'
export * from './ast'
