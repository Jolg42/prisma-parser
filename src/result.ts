export interface Result<T, Err extends Error> {
  isOk(): boolean
  isError(): boolean

  map<U>(mapper: (value: T) => U): Result<U, Err>
  flatMap<U>(mapper: (value: T) => Result<U, Err>): Result<U, Err>
  orElse(mapper: (error: Err) => Result<T, Err>): Result<T, Err>

  unwrap(): T
}

class ResultOk<T, Err extends Error> implements Result<T, Err> {
  private value: T
  constructor(value: T) {
    this.value = value
  }

  isOk(): boolean {
    return true
  }
  isError(): boolean {
    return false
  }
  map<U>(mapper: (value: T) => U): Result<U, Err> {
    return new ResultOk(mapper(this.value))
  }
  flatMap<U>(mapper: (value: T) => Result<U, Err>): Result<U, Err> {
    return mapper(this.value)
  }

  orElse(): Result<T, Err> {
    return ok(this.value)
  }

  unwrap(): T {
    return this.value
  }
}

class ResultErr<T, Err extends Error> implements Result<T, Err> {
  private error: Err
  constructor(error: Err) {
    this.error = error
  }

  isOk(): boolean {
    return false
  }

  isError(): boolean {
    return true
  }

  map<U>(): Result<U, Err> {
    return new ResultErr(this.error)
  }

  flatMap<U>(): Result<U, Err> {
    return new ResultErr(this.error)
  }

  orElse(mapper: (error: Err) => Result<T, Err>): Result<T, Err> {
    return mapper(this.error)
  }

  unwrap(): T {
    throw this.error
  }
}

export function ok<T, Err extends Error>(value: T): Result<T, Err> {
  return new ResultOk(value)
}

export function err<T, Err extends Error>(error: Err): Result<T, Err> {
  return new ResultErr(error)
}
