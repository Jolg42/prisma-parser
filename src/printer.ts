import { DocumentNode, ModelDefinitionNode, FieldDefinitionNode, Node, TypeNode, AttributeNode } from './ast'

const TAB_SIZE = 2

type PrintFunction<NodeType extends Node> = (node: NodeType, printerState: PrinterState) => void

function padBy(paddingSize: number, str: string): string {
  return str.padStart(str.length + paddingSize)
}

class PrinterState {
  private indentationLevel = 0
  private lines: string[] = []
  private currentLine = ''

  indent(): this {
    this.indentationLevel++
    return this
  }

  unindent(): this {
    if (this.indentationLevel === 0) {
      throw new Error('Can not unindent beyond 0')
    }
    this.indentationLevel--
    return this
  }

  write(string: string): this {
    this.currentLine += string
    return this
  }

  writeAligned(string: string, offsetFromStartOfTheLine: number): this {
    const paddingSize = Math.max(offsetFromStartOfTheLine - this.currentLine.length, 0)
    this.write(padBy(paddingSize, string))
    return this
  }

  newLine(): this {
    this.lines.push(padBy(this.indentationLevel * TAB_SIZE, this.currentLine))
    this.currentLine = ''
    return this
  }

  writeNewlineSeparated<NodeType extends Node>(nodes: NodeType[], printNode: PrintFunction<NodeType>): this {
    if (nodes.length === 0) {
      return this
    }
    printNode(nodes[0], this)
    for (let i = 1; i < nodes.length; i++) {
      this.newLine()
      printNode(nodes[i], this)
    }
    return this
  }

  getText(): string {
    if (this.currentLine) {
      this.newLine()
    }
    return this.lines.join('\n')
  }
}

export function print(document: DocumentNode): string {
  const printerState = new PrinterState()
  printDocument(document, printerState)
  return printerState.getText()
}

function printDocument(document: DocumentNode, printerState: PrinterState): void {
  printerState.writeNewlineSeparated(document.definitions, printDefinition)
  printerState.newLine()
}

function printDefinition(definition: ModelDefinitionNode, printerState: PrinterState): void {
  const typeOffset = getTypeAlignmentOffset(definition.fields)
  printerState
    .write('model ')
    .write(definition.name.identifier)
    .write(' {')
    .newLine()
    .indent()
    .writeNewlineSeparated(definition.fields, (field) => {
      printField(field, typeOffset, printerState)
    })
    .newLine()
    .unindent()
    .write('}')
    .newLine()
}

function getTypeAlignmentOffset(fields: FieldDefinitionNode[]): number {
  const longestFieldName = Math.max(...fields.map((field) => field.name.identifier.length))
  return longestFieldName + 2 // minimum ident is 2 spaces from field name
}

function printField(field: FieldDefinitionNode, typeOffset: number, printerState: PrinterState): void {
  printerState.write(field.name.identifier).write('  ')
  printType(field.type, typeOffset, printerState)

  for (const attribute of field.attributes) {
    printerState.write(' ')
    printAttribute(attribute, printerState)
  }
}

function printType(type: TypeNode, typeOffset: number, printerState: PrinterState): void {
  printerState.writeAligned(type.name.identifier, typeOffset)
  if (type.modifier === 'array') {
    printerState.write('[]')
  } else if (type.modifier === 'optional') {
    printerState.write('?')
  }
}

function printAttribute(attribute: AttributeNode, printerState: PrinterState): void {
  printerState.write(attribute.name)
}
