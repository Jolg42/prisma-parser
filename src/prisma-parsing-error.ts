import { createCodeFrame } from 'simple-code-frame'
import { Position } from './ast'

export class PrismaParsingError extends SyntaxError {
  position: Position
  source: string

  constructor(message: string, source: string, position: Position) {
    super(message)
    this.name = 'PrismaParsingError'
    this.source = source
    this.position = position
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, PrismaParsingError)
    }
  }

  toString(): string {
    // our line and column numbers are line-based, simple-code-frame ones are 0-based
    const frame = createCodeFrame(this.source, this.position.line - 1, this.position.column - 1, { colors: false })
    return `${this.name}: ${this.message}\n\n${frame}`
  }
}
