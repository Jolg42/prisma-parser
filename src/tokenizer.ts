import { Position } from './ast'
import { PrismaParsingError } from './prisma-parsing-error'

export type TokenType = 'identifier' | 'attribute' | 'string' | 'eof' | '{' | '}' | '(' | ')' | ',' | '[]' | '?' | '='

export type Token = {
  type: TokenType
  token: string
  start: Position
  end: Position
}

type TokenizationRule = {
  type: TokenType
  pattern: RegExp
}

const tokenizationRules: TokenizationRule[] = [
  { type: 'identifier', pattern: /[a-zA-Z][a-zA-Z0-9]*/y },
  { type: 'attribute', pattern: /@[a-zA-Z][a-zA-Z0-9]*/y },
  // eslint is wrong: \" is not an escape sequence, it is supposed to match
  // \ and " characters and \ does not need escaping in this case
  // eslint-disable-next-line no-useless-escape
  { type: 'string', pattern: /"([^"]|\")*"/y },
  { type: '{', pattern: /{/y },
  { type: '}', pattern: /}/y },
  { type: '(', pattern: /\(/y },
  { type: ')', pattern: /\)/y },
  { type: ',', pattern: /,/y },
  { type: '?', pattern: /\?/y },
  { type: '=', pattern: /=/y },
  { type: '[]', pattern: /\[]/y },
]

const SPACE_REGEX = /[\s\n\r]+/y

type LazyTokenizer = () => Token

function createLazyTokenizer(input: string): () => Token {
  let position: Position = {
    offset: 0,
    line: 1,
    column: 1,
  }

  const advancePosition = (byOffset: number): Position => {
    const offset = position.offset + byOffset
    let column = position.column
    let line = position.line
    for (let i = position.offset; i < offset; i++) {
      if (input[i] === '\n') {
        line++
        column = 1
      } else {
        column++
      }
    }

    return {
      offset,
      line,
      column,
    }
  }

  const consumeSpace = () => {
    SPACE_REGEX.lastIndex = position.offset
    const spaceMatch = SPACE_REGEX.exec(input)
    if (spaceMatch) {
      position = advancePosition(spaceMatch[0].length)
    }
  }

  return (): Token => {
    consumeSpace()
    if (position.offset === input.length) {
      return {
        type: 'eof',
        token: '',
        start: position,
        end: position,
      }
    }
    for (const rule of tokenizationRules) {
      rule.pattern.lastIndex = position.offset
      const match = rule.pattern.exec(input)
      if (!match) {
        continue
      }
      const token = match[0]
      const start = position
      const end = advancePosition(token.length)
      position = end

      return {
        type: rule.type,
        token,
        start,
        end,
      }
    }

    throw new PrismaParsingError('Unexpected token', input, position)
  }
}

export class TokenList {
  private currentTokenPointer = 0
  private computedTokens: Token[] = []
  private computeNextToken: LazyTokenizer

  constructor(input: string) {
    this.computeNextToken = createLazyTokenizer(input)
  }

  getPointer(): number {
    return this.currentTokenPointer
  }

  restorePointer(pointer: number): void {
    if (pointer >= this.computedTokens.length) {
      throw new Error('Attempt to set pointer to a not yet computed token')
    }

    this.currentTokenPointer = pointer
  }

  advance(): Token {
    const token = this.peek()
    this.currentTokenPointer++
    return token
  }

  peek(): Token {
    if (this.currentTokenPointer === this.computedTokens.length) {
      this.computedTokens.push(this.computeNextToken())
    }
    return this.computedTokens[this.currentTokenPointer]
  }
}
