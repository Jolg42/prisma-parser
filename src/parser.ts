import { Result, ok, err } from './result'
import { Token, TokenList, TokenType } from './tokenizer'
import { DocumentNode, TypeModifier, Position, ConfigType, FunctionCallNode, ExpressionNode } from './ast'
import { PrismaParsingError } from './prisma-parsing-error'

type ParsingResult<T> = Result<T, PrismaParsingError>
type RuleFunction<ReturnType> = (state: ParserState) => ParsingResult<ReturnType>

type TupleToUnion<TupleType extends unknown[]> = TupleType[number]

export function parse(source: string): DocumentNode {
  return document(new ParserState(source)).unwrap()
}

class ParserState {
  private tokenList: TokenList
  private source: string

  constructor(source: string) {
    this.source = source
    this.tokenList = new TokenList(source)
  }

  advance(): Token {
    return this.tokenList.advance()
  }

  peek(): Token {
    return this.tokenList.peek()
  }

  getPointer(): number {
    return this.tokenList.getPointer()
  }

  restorePointer(position: number): void {
    return this.tokenList.restorePointer(position)
  }

  error<T>(message: string, position: Position): ParsingResult<T> {
    return err(new PrismaParsingError(message, this.source, position))
  }
}

const document = mapRuleResult(
  () => takeUntil(definition, 'eof'),
  ([definitions, eof]): DocumentNode => ({
    kind: 'Document' as const,
    definitions,
    location: {
      start: definitions[0]?.location.start ?? 0,
      end: eof.end,
    },
  }),
)

const definition = defineRule(() => choice('Expected config or model definition', modelDefinition, configDefinition))

const modelDefinition = mapRuleResult(
  () =>
    sequence(
      keyword('model'),
      changeErrorMessage(identifier, 'Expected model name'),
      token('{'),
      takeUntil(fieldDefinition, '}'),
    ),
  ([modelKeyword, name, , [fields, closeBrace]]) => ({
    kind: 'ModelDefinition' as const,
    name,
    fields,
    location: {
      start: modelKeyword.start,
      end: closeBrace.end,
    },
  }),
)

const fieldDefinition = mapRuleResult(
  () => sequence(changeErrorMessage(identifier, 'Expected field name'), type, zeroOrMore(attribute)),
  ([name, type, attributes]) => {
    const end = attributes.length > 0 ? attributes[attributes.length - 1].location.end : type.location.end
    return {
      kind: 'FieldDefinition' as const,
      name,
      type,
      attributes,
      location: {
        start: name.location.start,
        end,
      },
    }
  },
)

const type = mapRuleResult(
  () => sequence(changeErrorMessage(identifier, 'Expected field type'), typeModifier),
  ([name, [modifier, modifierToken]]) => {
    return {
      kind: 'Type' as const,
      name,
      modifier,
      location: {
        start: name.location.start,
        end: modifierToken ? modifierToken.end : name.location.end,
      },
    }
  },
)

const typeModifier = mapRuleResult(
  () => optional(choice('Expected type modifier', token('?'), token('[]'))),
  (token): [TypeModifier, Token | null] => {
    if (!token) {
      return ['none', null]
    }
    if (token.token === '?') {
      return ['optional', token]
    }
    return ['array', token]
  },
)

const attribute = mapRuleResult(
  () => sequence(token('attribute'), optional(argumentsList)),
  ([token, argumentsList]) => {
    const [args, endToken] = argumentsList ?? [[], token]
    return {
      kind: 'Attribute' as const,
      name: token.token,
      arguments: args,
      location: {
        start: token.start,
        end: endToken.end,
      },
    }
  },
)

const configDefinition = mapRuleResult(
  () => sequence(configType, identifier, token('{'), takeUntil(configOption, '}')),
  ([configTypeToken, name, , [options, closingBrace]]) => ({
    kind: 'ConfigDefinition' as const,
    type: configTypeToken.token as ConfigType,
    name,
    options,
    location: {
      start: configTypeToken.start,
      end: closingBrace.end,
    },
  }),
)

const configType = defineRule(() =>
  choice('Expected generator or datasource keyword', keyword('datasource'), keyword('generator')),
)

const configOption = mapRuleResult(
  () => sequence(identifier, token('='), expression),
  ([key, , value]) => ({
    kind: 'ConfigOption' as const,
    key,
    value,
    location: {
      start: key.location.start,
      end: value.location.end,
    },
  }),
)

const expression = defineRule(() => choice('Expected expression', functionCall, stringLiteral, booleanLiteral))

const stringLiteral = mapRuleResult(
  () => token('string'),
  (token) => {
    const value = token.token.replace(/^"/, '').replace(/"$/, '').replace(/\\"/g, '"')
    return {
      kind: 'StringLiteral' as const,
      value,
      location: {
        start: token.start,
        end: token.end,
      },
    }
  },
)

const booleanLiteral = mapRuleResult(
  () => choice('Expected boolean value', keyword('true'), keyword('false')),
  (keywordToken) => ({
    kind: 'BooleanLiteral' as const,
    value: keywordToken.token === 'true' ? true : false,

    location: {
      start: keywordToken.start,
      end: keywordToken.end,
    },
  }),
)

const functionCall: RuleFunction<Required<FunctionCallNode>> = mapRuleResult(
  () => sequence(identifier, argumentsList),
  ([name, [args, closeBrace]]) => ({
    kind: 'FunctionCall' as const,
    name,
    arguments: args,
    location: {
      start: name.location.start,
      end: closeBrace.end,
    },
  }),
)

const argumentsList: RuleFunction<[ExpressionNode[], Token]> = mapRuleResult(
  () => sequence(token('('), zeroOrMoreSeparated(expression, token(',')), token(')')),
  ([, args, closingBrace]) => [args, closingBrace],
)

const identifier = mapRuleResult(
  () => token('identifier'),
  (token) => ({
    kind: 'Identifier' as const,
    identifier: token.token,
    location: {
      start: token.start,
      end: token.end,
    },
  }),
)

function sequence<Elements extends unknown[]>(
  ...ruleFunctions: { [K in keyof Elements]: RuleFunction<Elements[K]> }
): RuleFunction<Elements> {
  return defineRule(() => (state) => {
    let result = ok([] as unknown[])
    for (const rule of ruleFunctions) {
      result = result.flatMap((elements) => {
        return rule(state).map((nextValue) => [...elements, nextValue])
      })
    }

    return result as ParsingResult<Elements>
  })
}

function choice<Elements extends unknown[]>(
  defaultError: string,
  ...ruleFunctions: { [K in keyof Elements]: RuleFunction<Elements[K]> }
): RuleFunction<TupleToUnion<Elements>> {
  return defineRule(() => (state) => {
    let result = state.error(defaultError, state.peek().start)
    for (const rule of ruleFunctions) {
      result = result.orElse((previousError) => {
        return rule(state).orElse((nextError) => {
          // report longest partial match as an error of choice operator
          // chances are, that if one alternative matched more tokens that rule was users intention and this error will
          // be more relevant
          const errorToReport = nextError.position.offset > previousError.position.offset ? nextError : previousError
          return err(errorToReport)
        })
      })
    }

    return result as ParsingResult<TupleToUnion<Elements>>
  })
}

function optional<ReturnType>(rule: RuleFunction<ReturnType>): RuleFunction<ReturnType | null> {
  return defineRule(() => (state) => {
    return rule(state).orElse(() => ok(null))
  })
}

function takeUntil<ReturnType>(
  elementRule: RuleFunction<ReturnType>,
  tokenType: TokenType,
): RuleFunction<[ReturnType[], Token]> {
  return defineRule(() => (state) => {
    let result: ParsingResult<ReturnType[]> = ok([])
    while (state.peek().type !== tokenType && !result.isError()) {
      result = result.flatMap((elements) => {
        return elementRule(state).map((newElement) => [...elements, newElement])
      })
    }
    return result.map((values) => [values, state.advance()])
  })
}

function zeroOrMore<ReturnType>(elementRule: RuleFunction<ReturnType>): RuleFunction<ReturnType[]> {
  return defineRule(() => (state) => {
    const result: ReturnType[] = []
    let ruleResult = elementRule(state)
    while (ruleResult.isOk()) {
      result.push(ruleResult.unwrap())
      ruleResult = elementRule(state)
    }
    return ok(result)
  })
}

function zeroOrMoreSeparated<ReturnType>(
  elementRule: RuleFunction<ReturnType>,
  separatorRule: RuleFunction<unknown>,
): RuleFunction<ReturnType[]> {
  return defineRule(() => (state) => {
    const result: ReturnType[] = []
    let ruleResult = elementRule(state)
    while (ruleResult.isOk()) {
      result.push(ruleResult.unwrap())
      ruleResult = separatorRule(state).flatMap(() => elementRule(state))
    }
    return ok(result)
  })
}

const keyword = (keywordName: string): RuleFunction<Token> => {
  return defineRule(() => (state) => {
    return identiferToken(state)
      .orElse<Token>((error) => state.error(`Expected keyword ${keywordName}`, error.position))
      .flatMap((token) => {
        if (token.token !== keywordName) {
          return state.error(`Expected keyword ${keywordName}`, token.start)
        }
        return ok(token)
      })
  })
}

const identiferToken = token('identifier')

function token(tokenType: TokenType): RuleFunction<Token> {
  return defineRule(() => (state) => {
    const token = state.advance()
    if (token.type === tokenType) {
      return ok(token)
    }
    return state.error(`Expected ${tokenType}`, token.start)
  })
}

function mapRuleResult<T, U>(ruleFactory: () => RuleFunction<T>, mapper: (result: T) => U): RuleFunction<U> {
  return defineRule(() => {
    const ruleFn = ruleFactory()
    return (state) => ruleFn(state).map(mapper)
  })
}

function changeErrorMessage<T>(ruleFn: RuleFunction<T>, message: string): RuleFunction<T> {
  return (state) => {
    return ruleFn(state).orElse((error) => state.error(message, error.position))
  }
}

function defineRule<ReturnType>(ruleFactory: () => RuleFunction<ReturnType>): RuleFunction<ReturnType> {
  let ruleFn: RuleFunction<ReturnType> | null = null
  return (state) => {
    if (!ruleFn) {
      ruleFn = ruleFactory()
    }
    const position = state.getPointer()
    const result = ruleFn(state)
    if (result.isError()) {
      state.restorePointer(position)
    }
    return result
  }
}
