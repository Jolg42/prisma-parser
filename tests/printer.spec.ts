import { print } from '../src/printer'

test('prints empty document', () => {
  expect(
    print({
      kind: 'Document',
      definitions: [],
    }),
  ).toMatchSnapshot()
})

test('prints empty model', () => {
  expect(
    print({
      kind: 'Document',
      definitions: [
        {
          kind: 'ModelDefinition',
          name: { kind: 'Identifier', identifier: 'Model' },
          fields: [],
        },
      ],
    }),
  ).toMatchSnapshot()
})

test('prints field', () => {
  expect(
    print({
      kind: 'Document',
      definitions: [
        {
          kind: 'ModelDefinition',
          name: { kind: 'Identifier', identifier: 'Model' },
          fields: [
            {
              kind: 'FieldDefinition',
              name: { kind: 'Identifier', identifier: 'field' },
              type: {
                kind: 'Type',
                name: { kind: 'Identifier', identifier: 'Int' },
                modifier: 'none',
              },
              attributes: [],
            },
          ],
        },
      ],
    }),
  ).toMatchSnapshot()
})

test('prints field with array type', () => {
  expect(
    print({
      kind: 'Document',
      definitions: [
        {
          kind: 'ModelDefinition',
          name: { kind: 'Identifier', identifier: 'Model' },
          fields: [
            {
              kind: 'FieldDefinition',
              name: { kind: 'Identifier', identifier: 'array' },
              type: {
                kind: 'Type',
                name: { kind: 'Identifier', identifier: 'Int' },
                modifier: 'array',
              },
              attributes: [],
            },
          ],
        },
      ],
    }),
  ).toMatchSnapshot()
})

test('prints field with optional type', () => {
  expect(
    print({
      kind: 'Document',
      definitions: [
        {
          kind: 'ModelDefinition',
          name: { kind: 'Identifier', identifier: 'Model' },
          fields: [
            {
              kind: 'FieldDefinition',
              name: { kind: 'Identifier', identifier: 'array' },
              type: {
                kind: 'Type',
                name: { kind: 'Identifier', identifier: 'Int' },
                modifier: 'optional',
              },
              attributes: [],
            },
          ],
        },
      ],
    }),
  ).toMatchSnapshot()
})

test('prints field attribute', () => {
  expect(
    print({
      kind: 'Document',
      definitions: [
        {
          kind: 'ModelDefinition',
          name: { kind: 'Identifier', identifier: 'Model' },
          fields: [
            {
              kind: 'FieldDefinition',
              name: { kind: 'Identifier', identifier: 'id' },
              type: {
                kind: 'Type',
                name: { kind: 'Identifier', identifier: 'Int' },
                modifier: 'none',
              },
              attributes: [
                {
                  kind: 'Attribute',
                  name: '@id',
                },
              ],
            },
          ],
        },
      ],
    }),
  ).toMatchSnapshot()
})

test('prints multiple attributes', () => {
  expect(
    print({
      kind: 'Document',
      definitions: [
        {
          kind: 'ModelDefinition',
          name: { kind: 'Identifier', identifier: 'Model' },
          fields: [
            {
              kind: 'FieldDefinition',
              name: { kind: 'Identifier', identifier: 'field' },
              type: {
                kind: 'Type',
                name: { kind: 'Identifier', identifier: 'Int' },
                modifier: 'none',
              },
              attributes: [
                {
                  kind: 'Attribute',
                  name: '@first',
                },

                {
                  kind: 'Attribute',
                  name: '@second',
                },
              ],
            },
          ],
        },
      ],
    }),
  ).toMatchSnapshot()
})

test('prints multiple fields while aligning the types', () => {
  expect(
    print({
      kind: 'Document',
      definitions: [
        {
          kind: 'ModelDefinition',
          name: { kind: 'Identifier', identifier: 'Model' },
          fields: [
            {
              kind: 'FieldDefinition',
              name: { kind: 'Identifier', identifier: 'short' },
              type: {
                kind: 'Type',
                name: { kind: 'Identifier', identifier: 'Int' },
                modifier: 'none',
              },
              attributes: [],
            },

            {
              kind: 'FieldDefinition',
              name: { kind: 'Identifier', identifier: 'loooong' },
              type: {
                kind: 'Type',
                name: { kind: 'Identifier', identifier: 'String' },
                modifier: 'none',
              },
              attributes: [],
            },
          ],
        },
      ],
    }),
  ).toMatchSnapshot()
})
