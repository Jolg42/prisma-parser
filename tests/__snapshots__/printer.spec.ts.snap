// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prints empty document 1`] = `""`;

exports[`prints empty model 1`] = `
"model Model {
  
}
"
`;

exports[`prints field 1`] = `
"model Model {
  field  Int
}
"
`;

exports[`prints field attribute 1`] = `
"model Model {
  id  Int @id
}
"
`;

exports[`prints field with array type 1`] = `
"model Model {
  array  Int[]
}
"
`;

exports[`prints field with optional type 1`] = `
"model Model {
  array  Int?
}
"
`;

exports[`prints multiple attributes 1`] = `
"model Model {
  field  Int @first @second
}
"
`;

exports[`prints multiple fields while aligning the types 1`] = `
"model Model {
  short    Int
  loooong  String
}
"
`;
